kratos/python_scripts/analysis_stage.py

```diff
@@ -1,305 +1,384 @@
+            for process in self._GetListOfProcesses():
+                process.ExecuteBeforeOutputStep()
+
+            for output_process in self._GetListOfOutputProcesses():
+                if output_process.IsOutputStep():
+                    output_process.PrintOutput()
+
+            for process in self._GetListOfProcesses():
+                process.ExecuteAfterOutputStep()
+
+    def Check(self):
+        """This function checks the AnalysisStage
+        """
+        self._GetSolver().Check()
+        for process in self._GetListOfProcesses():
+            process.Check()
+
+    def Clear(self):
+        """This function clears the AnalysisStage
+        """
+        self._GetSolver().Clear()
+
+    def ModifyInitialProperties(self):
+        """this is the place to eventually modify material properties in the stage """
+        pass
+
+    def ModifyInitialGeometry(self):
+        """this is the place to eventually modify geometry (for example moving nodes) in the stage """
+        pass
+
+    def ModifyAfterSolverInitialize(self):
+        """this is the place to eventually do any modification that requires the solver to be initialized """
+        pass
+
+    def ApplyBoundaryConditions(self):
+        """here the boundary conditions is applied, by calling the InitializeSolutionStep function of the processes"""
+
+        for process in self._GetListOfProcesses():
+            process.ExecuteInitializeSolutionStep()
+
+        #other operations as needed
+
+    def ChangeMaterialProperties(self):
+        """this function is where the user could change material parameters as a part of the solution step """
+        pass
+
+    def _GetSolver(self):
+        if not hasattr(self, '_solver'):
+            self._solver = self._CreateSolver()
+        return self._solver
+
+    def _CreateSolver(self):
+        """Create the solver
+        """
+        raise Exception("Creation of the solver must be implemented in the derived class.")
+
+    ### Modelers
+    def _SetupGeometryModelAndModelParts(self):
+        ## Initialize Modelers
+        self._list_of_modelers = self._CreateModelers(self._GetOrderOfModelers())
+
+        # Import or generate geometry models from external input.
+        for modeler in self._GetListOfModelers():
+            modeler.SetupGeometryModel(self.model)
+
+        # Prepare or update the geometry model_part.
+        for modeler in self._GetListOfModelers():
+            modeler.PrepareGeometryModel(self.model)
+
+        # Convert the geometry model or import analysis suitable models.
+        for modeler in self._GetListOfModelers():
+            modeler.SetupModelPart(self.model)
+
+    ### Modelers
+    def _GetListOfModelers(self):
+        """ This function returns the list of modelers
+        """
+        if not hasattr(self, '_list_of_modelers'):
+            raise Exception("The list of modelers was not yet created!")
+        return self._list_of_modelers
+
+    def _GetOrderOfModelers(self):

```

Comment:
then you also won't need this function which is a real pain to use/understand/debug